### PR TITLE
obs-ffmpeg: Fix crash when retrying init encoder for NVENC

### LINKED ISSUE #6911: OBS crashes when selecting lossless in "NVENC (new)"
### Operating System Info

Windows 10

### Other OS

_No response_

### OBS Studio Version

Git

### OBS Studio Version (Other)

_No response_

### OBS Studio Log URL

https://obsproject.com/logs/Jr3Taws1zttuaUZf

### OBS Studio Crash Log URL

https://obsproject.com/logs/Kf3GKDsiktJ53zP0

### Expected Behavior

OBS starts output without crashing

### Current Behavior

OBS crashes.

### Steps to Reproduce

1. download Nightly Build
2. Open Settings -> Output
3. Set output mode to "Advanced
4. Open the Recording tab
5. Set to "NVIDIA NVENC H.264 (new)
or "NVIDIA NVENC HEVC (new)".
6. Change "Rate Control" to "Lossless".
7. Start Recording (or Start Replay Buffer)
8. OBS crashes.

### Anything else we should know?

This does not occur in 27.2.4.

If "Psycho Visual Tuning" is disabled, the crash will not occur. (Why?)

I'm using translation. Sorry if the English is not correct.

### PR DESCRIPTION
**Pull request by tuduweb**
### Description

1. Correctly assign argument psycho_aq to false when retrying again without Psycho Visual Tuning.
2. Reinit encode session before retrying init encoder to fix crash
3. Code cleanup

### Motivation and Context

1. notice that the argument `psycho_aq` was assigned unexpectedly when debugging
2. fixes #6911 
3. code cleanup mentioned by @jpark37 

the diff between 28 and 27.2.4(doesnt crash in this version mentioned by issue) https://github.com/obsproject/obs-studio/commit/699e89d8596e8f96c3d7d4448998ca73d2efac87
move the retrying code segments, and then in this way pass the `nvenc_destroy(enc)` and `init_session(enc)` procedure.

https://github.com/obsproject/obs-studio/blob/1444ead8123a3947ea12f1f515afd5741ee5936e/plugins/obs-ffmpeg/jim-nvenc.c#L689-L696

https://github.com/obsproject/obs-studio/blob/1444ead8123a3947ea12f1f515afd5741ee5936e/plugins/obs-ffmpeg/jim-nvenc.c#L629-L664

### How Has This Been Tested?

Win10 with Nvidia GTX 1050 Ti max-Q Design (Laptop version), Thinkpad X1e 1gen

retry these steps in *Steps to Reproduce*
with settings `X264(new)`/`HEVC(new)`, `Rescale Output(disable)` and `Lossless`

### Types of changes

Bug fix (non-breaking change which fixes an issue)
Code cleanup (non-breaking change which makes code smaller or more readable)
### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

File: plugins/obs-ffmpeg/jim-nvenc.c [hunk 1]
```diff
@@ -997,8 +1005,9 @@ static bool init_encoder(struct nvenc_data *enc, bool hevc,
 
 		blog(LOG_WARNING, "[jim-nvenc] init_specific_encoder failed, "
 				  "trying again without Psycho Visual Tuning");
+		reinit_session(enc);
 		if (!init_specific_encoder(enc, hevc, settings, encoder, bf,
-					   psycho_aq)) {
+					   false)) {
```
Comment (theHamsta) [inline plugins/obs-ffmpeg/jim-nvenc.c:1010]: can confirm that I had the same problem. Current versions of the SDK also allow for testing for `psycho_aq` support beforehand

Comment (jpark37) [reply plugins/obs-ffmpeg/jim-nvenc.c:1010]: What would that test look like?

File: plugins/obs-ffmpeg/jim-nvenc.c [hunk 2]
```diff
@@ -997,8 +1005,9 @@ static bool init_encoder(struct nvenc_data *enc, bool hevc,
 
 		blog(LOG_WARNING, "[jim-nvenc] init_specific_encoder failed, "
 				  "trying again without Psycho Visual Tuning");
+		reinit_session(enc);
```
Comment (jpark37) [inline plugins/obs-ffmpeg/jim-nvenc.c:1008]: Can you do this instead?
```
blog(LOG_WARNING, "[jim-nvenc] init_specific_encoder failed, "
		  "trying again without Psycho Visual Tuning");
nv.nvEncDestroyEncoder(enc->session);
enc->session = NULL;
if (!init_session(enc) ||
    !init_specific_encoder(enc, hevc, settings, encoder, bf,
			   false)) {
	return false;
}
```

**Review by jpark37 — CHANGES_REQUESTED** [review]
Can you work NVENC into the PR/commit titles? obs-ffmpeg has many encoders, so it's nice to see at a glance which one.

Also, there's an unused variable in init_specific_encoder. Can you remove that while you're here?

Thanks.

**Comment by energizerfellow** [comment]
The original backing off of psycho visual tuning disabled both spatial and temporal AQ when it only needed to disable temporal. Having functional spatial AQ on the temporal-less Kepler would be a nice quality bump for those still on Kepler.

https://docs.nvidia.com/video-technologies/video-codec-sdk/nvenc-application-note/#nvenc-capabilities

File: plugins/obs-ffmpeg/jim-nvenc.c (hunk 2 above)
Comment (tuduweb) [reply plugins/obs-ffmpeg/jim-nvenc.c:1008]: yea, it seem like that I have ignored the init_session return value here.
thx, I will make this changes soon.

Comment (jpark37) [reply plugins/obs-ffmpeg/jim-nvenc.c:1008]: I'm also not a fan of stubby little helper functions if they're not called more than once. I'd rather see the logic directly in context.

Thanks for the bugfix.

File: plugins/obs-ffmpeg/jim-nvenc.c (hunk 1 above)
Comment (tuduweb) [reply plugins/obs-ffmpeg/jim-nvenc.c:1003]: hi, do you mean `nv_get_cap_h264(enc, NV_ENC_CAPS_SUPPORT_LOSSLESS_ENCODE)` for testing or something else?

> Can you work NVENC into the PR/commit titles? obs-ffmpeg has many encoders, so it's nice to see at a glance which one.
> 
> Also, there's an unused variable in init_specific_encoder. Can you remove that while you're here?
> 
> Thanks.

it done.
thanks for your revsion.

**Comment by tuduweb** [comment]
> Can you work NVENC into the PR/commit titles? obs-ffmpeg has many encoders, so it's nice to see at a glance which one.
> 
> Also, there's an unused variable in init_specific_encoder. Can you remove that while you're here?
> 
> Thanks.

it done.
thanks for your revsion.

File: plugins/obs-ffmpeg/jim-nvenc.c (hunk 2 above)
Comment (tuduweb) [reply plugins/obs-ffmpeg/jim-nvenc.c]: Yes, I agree with you.
Maybe I think using `reinit` help function can help understanding and make the code more readable though only used once.

File: plugins/obs-ffmpeg/jim-nvenc.c [hunk 3]
```diff
@@ -949,8 +949,7 @@ static bool init_textures(struct nvenc_data *enc)
 static void nvenc_destroy(void *data);
 
 static bool init_specific_encoder(struct nvenc_data *enc, bool hevc,
-				  obs_data_t *settings, obs_encoder_t *encoder,
-				  int bf, bool psycho_aq)
+				  obs_data_t *settings, int bf, bool psycho_aq)
 {
 	bool init = false;
```
Comment (jpark37) [inline plugins/obs-ffmpeg/jim-nvenc.c:954]: I mean this variable. I didn't even see the other one, lol. Good to get rid of both.

File: plugins/obs-ffmpeg/jim-nvenc.c (hunk 1 above)
Comment (jpark37) [reply plugins/obs-ffmpeg/jim-nvenc.c]: Resolving this because it's not an action item.

**Comment by jpark37** [comment]
> Current versions of the SDK also allow for testing for psycho_aq support beforehand

@theHamsta Which function would this be?

File: plugins/obs-ffmpeg/jim-nvenc.c (hunk 1 above)
Comment (tuduweb) [reply plugins/obs-ffmpeg/jim-nvenc.c]: do you mean `nv_get_cap_h264(enc, NV_ENC_CAPS_SUPPORT_LOSSLESS_ENCODE)` for testing support or something else?
could you please give more information.
thx. @theHamsta

File: plugins/obs-ffmpeg/jim-nvenc.c (hunk 3 above)
Comment (tuduweb) [reply plugins/obs-ffmpeg/jim-nvenc.c:954]: yea, i got it.
I just thought the variable was reserved for something by you when I was fixing.

Comment (jpark37) [reply plugins/obs-ffmpeg/jim-nvenc.c:954]: You removed `bool init = false;`? I still see it.

Comment (tuduweb) [reply plugins/obs-ffmpeg/jim-nvenc.c:954]: no. I re-push it now.

File: plugins/obs-ffmpeg/jim-nvenc.c (hunk 1 above)
Comment (theHamsta) [reply plugins/obs-ffmpeg/jim-nvenc.c]: @jpark37 yes, this nvEncGetEncodeCaps/NV_ENC_CAPS_SUPPORT_TEMPORAL_AQ is what I meant. It's already checked in `init_encoder_h264` and ignored if not supported

**Review by jpark37 — APPROVED** [review]

File: plugins/obs-ffmpeg/jim-nvenc.c (hunk 1 above)
Comment (jpark37) [reply plugins/obs-ffmpeg/jim-nvenc.c]: K, thanks. I think I'm happy with where this PR is for the time being.

**Pull request merged**